Mathlib/Data/Set/Image.lean

```diff
@@ -329,6 +329,17 @@ theorem image_subset {a b : Set α} (f : α → β) (h : a ⊆ b) : f '' a ⊆ f
   exact fun x => fun ⟨w, h1, h2⟩ => ⟨w, h h1, h2⟩
 #align set.image_subset Set.image_subset
 
+theorem subset_image_iff {t : Set β} :
+    t ⊆ f '' s ↔ ∃ u, u ⊆ s ∧ f '' u = t := by
```

Comment:
I think you should be able to use [Set.image_preimage_eq_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Image.html#Set.image_preimage_eq_iff) or one of the lemmas around, either in the proof of this or as a replacement for this.

Reply:
The obvious (to me) way to use existing lemmas needs [Set.range_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Function.html#Set.range_restrict).

Granting that we can have:
```
theorem subset_image_iff {f : α → β} {s : Set α} {t : Set β} :
    t ⊆ f '' s ↔ ∃ u, u ⊆ s ∧ f '' u = t := by
  rw [← range_restrict, subset_range_iff_exists_image_eq]
  refine ⟨fun ⟨u, hu⟩ ↦ ⟨(↑) '' u, ?_, ?_⟩, fun ⟨u, hu, hu'⟩ ↦ ⟨((↑) : s → α) ⁻¹' u, ?_⟩⟩ <;> aesop
```
The `aesop` is also not necessarily but it saves some annoying fiddling with `ext` and `simp`.

Do you prefer this?

Reply:
I was thinking about something less tricky, but indeed while proving it I realized there was one part I didn't think about. 
If you put this after [image_inter_preimage](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Image.html#Set.image_inter_preimage), you can do:
```lean
theorem subset_image_iff {t : Set β} :
    t ⊆ f '' s ↔ ∃ u, u ⊆ s ∧ f '' u = t := by
  refine ⟨fun h ↦ ⟨f ⁻¹' t ∩ s, inter_subset_right _ _, ?_⟩,
    fun ⟨u, hu, hu'⟩ ↦ hu'.symm ▸ image_mono hu⟩
  rwa [Set.image_preimage_inter, inter_eq_left]
```

Reply:
Neat!